Create torch.distributed.shard package.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71742

We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

pritamdamania committed:
Create torch.distributed.shard package.

We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good, thanks for doing this!

fduwjj commented on torch/distributed/shard/api.py:
```diff
@@ -0,0 +1,145 @@
+    ShardedTensor,
+)
+
+def shard_parameter(
```
Like this one. Then we can have reshard API also under this folder.

fduwjj reviewed (approved):
LGTM. Thanks for making this look cleaner and organized.

pritamdamania committed:
Update on "Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Create torch.distributed.shard package."


We have many sharding components currently:
torch.distributed._sharded_tensor, torch.distributed._sharding_spec,
torch.distributed._sharded_optimizer and more coming.

As a result, organizing all of this under the `torch.distributed.shard`
package. For BC reasons, I'm still keeping the old packages and have them just
reference the new package.

Differential Revision: [D33755913](https://our.internmc.facebook.com/intern/diff/D33755913/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33755913/)!

[ghstack-poisoned]
